src/main/kotlin/bsu/cc/parser/XlsxParser.kt

```diff
@@ -0,0 +1,152 @@
+        val rowCell = indexedRowMap[index]
+        if(colName == null) {
+            throw IllegalArgumentException("Accessing column index out of bounds of headers")
+        }
+        if(rowCell == null) {
+            throw IllegalStateException("Key not in map") //This should never happen, but needed for smartcast
+        }
+        if(!output.containsKey(colName)) {
+            output[colName] = rowCell
+        } else if(!ignoreDuplicateColumns) {
+            throw IllegalArgumentException("Column map contains duplicate column names")
+        }
+    }
+    return output
+}
+
+fun sheetHeaderMap(sheet: Sheet): Map<Int, String> {
+    val output = HashMap<Int, String>()
+    val headerRow = sheet.first()
+    //Can't use foreach iterator because it skips empty columns, which causes column indices to be wrong
+    (0 until headerRow.lastCellNum).forEach{index ->
+        if(headerRow.getCell(index) != null) {
+            val cell = headerRow.getCell(index)
+            if(cell != null) {
+                if(cell.cellType != CellType.STRING) {
+                    throw IllegalArgumentException("Sheet has non-string in header")
+                }
+                output[index] = cell.stringCellValue
+            }
+        }
+    }
+    return output
+}
+
+fun <T> rowToDataClass(
+        row: Row,
+        dataProducer: (rowMap: Map<Int, Cell>) -> T
+): T {
+    return dataProducer(rowToCellMap(row))
+}
+
+inline fun <reified T : Any> getFromCellOrThrow(cell: Cell?): T {
+    if(cell == null) {
+        throw IllegalArgumentException("Cell is null")
+    }
+
+    val expectedCellType = typeToCellType(T::class.java)
+    if(expectedCellType != cell.cellType) {
+        throw IllegalArgumentException("Attempting to get $expectedCellType from cell of type ${cell.cellType}")
+    }
+
+    return when(cell.cellType) {
+        CellType.STRING -> cell.stringCellValue as T
+        CellType.NUMERIC -> if(T::class.java == Date::class.java) cell.dateCellValue as T else cell.numericCellValue as T
+        else -> throw IllegalArgumentException("Attempting to get cell value for not string/numeric")
+    }
+}
+
+inline fun <reified T : Any> getFromCellOrDefault(cell: Cell?, default: T): T {
+    if(cell == null) {
+        return default
+    }
+
+    val expectedCellType = typeToCellType(T::class.java)
+    if(expectedCellType != cell.cellType) {
+        return default
+    }
+
+    return when(cell.cellType) {
+        CellType.STRING -> cell.stringCellValue as T
+        CellType.NUMERIC -> if(T::class.java == Date::class.java) cell.dateCellValue as T else cell.numericCellValue as T
+        else -> default
+    }
+}
+
+fun typeToCellType(javaType: Type): CellType {
+    return when(javaType) {
+        Int::class.java, Integer::class.java  -> throw IllegalArgumentException("Numeric values in Apache POI are treated as doubles, Ints are not supported")
+        java.lang.Double::class.java -> CellType.NUMERIC
```

Comment:
I really wish kotlin provided a nicer way to deal with box primitive class objects